### PR TITLE
wip/6.0 Fix thread safety issue in QueryInterpretationCacheStandardImpl

### PR DESCRIPTION
**Pull request by NathanQingyangXu**
As explained in https://dzone.com/articles/concurrenthashmap-isnt-always-enough, **ConcurrentHashMap**'s **computeIfAbsent**() method enjoys the benefit of atomicity (for the same key only one thread is running while others are blocked) so we can avoid many complicated race conditions.

It is a pity that the **BoundedConcurrentHashMap** class doesn't provide atomic overridden '**computeIfAbsent**()' method as compared to the **ConcurrentHashMap** from JDK. However, the code changes in this PR still enjoys the fact that internally '**putIfAbsent**()' (which is atomic) is used for unknown key so we can avoid populating cache multiple times (the earliest one wins).

This PR fixes the atomicity issue of 'computeIfAbsent()' for **BoundedConcurrentHashMap** as well.

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/query/internal/QueryInterpretationCacheStandardImpl.java [hunk 1]
```diff
@@ -77,14 +77,7 @@ public SelectQueryPlan resolveSelectQueryPlan(
...
 
-		final SelectQueryPlan cached = (SelectQueryPlan) queryPlanCache.get( key );
-		if ( cached != null ) {
-			return cached;
-		}
-
-		final SelectQueryPlan plan = creator.get();
-		queryPlanCache.put( key, plan );
-		return plan;
+		return (SelectQueryPlan) queryPlanCache.computeIfAbsent( key, k -> creator.get() );
```
Comment (NathanQingyangXu) [inline hibernate-core/src/main/java/org/hibernate/query/internal/QueryInterpretationCacheStandardImpl.java:80]: internally the following default implementation of 'computeIfAbsent' from ConcurrentMap will be run. It is a pity we cannot enjoy the atomicity in ConcurrentHashMap in JDK, but the usage of 'putIfAbsent()' is definitively desirable:
Screen Shot 2020-06-13 at 2 46 19 PM

**Comment by NathanQingyangXu** [comment]
I would try to fix the 'computeIfAbsent()' atomicity issue in BoundedConcurrnetHashMap. For this reason, I close this PR for the time being

**Comment by NathanQingyangXu** [comment]
I need more time to solve the testing failure, so I close this PR for the time being.

**Comment by Sanne** [comment]
hi @NathanQingyangXu , are you familiar with the notion of benign data races?

Please don't fix this: while there is a possibility of running a query twice, this is very unlikely. But having the lock would introduce a very severe limitation to horizontal scalability - so the tradeoff isn't good.

**Comment by NathanQingyangXu** [comment]
okay, I have closed the PR if that is not a valid concern.

**Comment by NathanQingyangXu** [comment]
that being said, the fix won't add overhead for we have used concurrent
map. We simply replace two locks with one. the real downside is code
complexity and maintainability. Seems Guava cache shares much with our
BoundedConcurrentHashmap, and it has atomic support for 'compute' map
methods.

**Comment by Sanne** [comment]
The "atomic" implementation of JDK's ConcurrentHashMap requries a blocking lock on the segment, which would severly limit scalability.

I have not checked the implementation from Guava.. maybe there's something smarter there, but please check for that with caution: it's better to occasionally miss the cache than to introduce blocking operations while accessing the cache.

**Comment by NathanQingyangXu** [comment]
the cache we are using is BoundedConcurrentHashmap, it was copied from
hibernate-infinity and is based on some early version of ConcurrentHashmap.
We have paid the cost already.

**Comment by Sanne** [comment]
> the cache we are using is BoundedConcurrentHashmap, it was copied from
hibernate-infinity and is based on some early version of ConcurrentHashmap.

It was based on the library that preceded Caffeine - forgot the name now.

> We have paid the cost already.

I'm not understanding, what cost are you referring to?

Thanks!

**Comment by NathanQingyangXu** [comment]
forget about it. I misunderstood something. Thanks for the clarification.

**Pull request closed without merging**